File: src/01/concurrent.py
import threading
import time

def counting(name, n):
    for i in range(n):
        print('thread {}: {}'.format(name, i))
        time.sleep(0.1)

def sayhello(name, n):
    for i in range(n, -1, -1):
        print('thread {}: hellooooo'.format(name))
        time.sleep(0.1)

N = 4
threads = []
for i in range(N):
    cmd = counting if i % 2 == 0 else sayhello
    thread = threading.Thread(target=cmd, args=(i, 10))
    threads.append(thread)
    thread.start()

for t in threads:
    t.join()

File: src/01/concurrent.go
package main;

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

func counting(name string, n int) {
	d, _ := time.ParseDuration(fmt.Sprintf("%dms", rand.Intn(1000)))
	for i := 0; i < n; i++ {
		fmt.Printf("Thead %s: %d ", name, i)
		time.Sleep(d)
	}
}

func main() {
	names := []string{"budi", "agus", "iwan", "susi"}
	var wg sync.WaitGroup

	wg.Add(len(names))
	for _, name := range names {
		go func(name string) {
			defer wg.Done()
			counting(name, 10)
		}(name)
	}

	wg.Wait()
}
